docs(deepagents): add filesystem permissions page (#3534)
langchain-ai/docs: src/oss/deepagents/permissions.mdx

---
title: Permissions
description: Control filesystem access with declarative permission rules for Deep Agents
---

> Control which files and directories an agent can read or write using declarative permission rules. Pass a list of rules to `permissions=` and the agent's built-in filesystem tools respect them automatically, without writing custom [backend policy hooks](/oss/deepagents/backends#add-policy-hooks).

Reviewer: I think we don't need to mention the previous workaround?
```suggestion
Control which files and directories an agent can read or write to using declarative permission rules. Pass a list of rules to `permissions=` and the agent's built-in filesystem tools respect them.
```

> Control which files and directories an agent can read or write using declarative permission rules. Pass a list of rules to `permissions=` and the agent's built-in filesystem tools respect them automatically, without writing custom [backend policy hooks](/oss/deepagents/backends#add-policy-hooks).

Reviewer: Add the warning text here as normal text to reiterate it, I think that's enough. 
```suggestion
Control which files and directories an agent can read or write using declarative permission rules. Pass a list of rules to `permissions=` and the agent's built-in filesystem tools respect them.

Permissions only apply to the built-in filesystem tools (`ls`, `read_file`, `glob`, `grep`, `write_file`, `edit_file`). Custom tools and MCP tools that access the filesystem are not covered.
```

)

# Works: permissions are scoped to the /memories/ route
agent = create_deep_agent(
    model=model,
    backend=composite,
    permissions=[
        FilesystemPermission(
            operations=["write"],
            paths=["/memories/**"],
            mode="deny",
        ),
    ],
)
```

Permissions that include paths outside any route raise `NotImplementedError`:

```python
# Raises NotImplementedError: /workspace/** hits the sandbox default
agent = create_deep_agent(
    model=model,
    backend=composite,
    permissions=[
        FilesystemPermission(
            operations=["write"],
            paths=["/workspace/**"],
            mode="deny",
        ),
    ],
)

# Also raises: /** covers both routes and the default
agent = create_deep_agent(
    model=model,
    backend=composite,
    permissions=[
        FilesystemPermission(
            operations=["read"],
            paths=["/**"],
            mode="deny",
        ),
    ],
)
```

:::

## Limitations

<Warning>
Permissions only apply to the built-in filesystem tools (`ls`, `read_file`, `glob`, `grep`, `write_file`, `edit_file`). Custom tools and MCP tools that access the filesystem are not covered.

> </Warning>

Reviewer: ```suggestion
```

<Warning>
Sandbox backends support arbitrary command execution, which can bypass path-based restrictions. Using permissions with a sandbox backend raises `NotImplementedError` unless all permission paths are scoped to [composite backend](/oss/deepagents/backends) routes.

> </Warning>

Reviewer: - Generally we don't do multiple warning callouts in a row. I think in this case we need these warnings also further above. At the bottom of the page is too easy to miss.

> </Warning>

Reviewer: Is this a limitation or is this desired? Either way I think there's no danger so this doesn't need to be a warning does it? I think if you want to point out Permissions do not apply to sandbox backends you can put this in the top or with the sandbox examples or both

```suggestion
```
